django_netjsonconfig/static/django-netjsonconfig/js/widget.js

```diff
@@ -67,13 +67,44 @@
         return editor;
     };
 
+    var evaluateVars = function(data, context){
+        var pattern = new RegExp('\{\{.*?\}\}');
+                if (typeof data === 'object'){
+                    for (let [key, value] of Object.entries(data)){
+                        data[key] = evaluateVars(value, context);
+                    }
+                }
+                if (typeof data === 'string') {
+                    var found_vars = data.match(pattern);
+                    if (found_vars != null){
+                        found_vars.forEach(function(element) {
+                            element = element.replace(/^\{\{\s+|\s+\}\}$|^\{\{|\}\}$/g, '');
+                            if (element in context){
+                                data = data.replace(pattern, context[element]);
+                            }
+                        });
+                    }
+                }
+        return data;
+    };
+
     // returns true if JSON is well formed
     // and valid according to its schema
     var isValidJson = function(advanced){
-        var valid;
-        try{
-            valid = advanced.validateSchema(advanced.get());
-        }catch (e){
+        var valid,
+            default_values = {},
+            data,
+            cleanedData;
+        try {
+            try {
+                default_values = JSON.parse($('#id_default_values').val());
+            } catch(e) {
+                default_values = JSON.parse($('#id_config-0-context').val());
```

Comment:
unfortunately this is not enough. Because the context of a device can have many variables, not only the ones defined in the context field. We need to make something more complex (get the context from the backend via an API endpoint), but we don't have to do it now. Please also add an issue for this task in the project board.

Reply:
@nemesisdesign please i have not understood this point. Please how can we get the value of context from the backend API while we are typing.

Reply:
we have to add a new view to get the full context of from the backend.

We can get the full context of a device, while for templates and VPNs we can design some placeholder value that would pass the validation.

This API endpoint would be called when the editor is loaded. Is it clear? If not we can discsuss.

Reply:
Please @nemesisdesign  I have not still understood. are we saying that device can already have some predefined context values and we need them here to pass the validation? Also, when you talked of placeholder values, do you mean we set values in the code? if so, then how will the user entered values be validated against the schema for the set variables?

Reply:
Look here: https://github.com/openwisp/django-netjsonconfig/blob/master/django_netjsonconfig/base/config.py#L94-L111

And here: https://github.com/openwisp/django-netjsonconfig#netjsonconfig_context

Each device gets some variables automatically. We can define some placeholders for those kind of variables that can be used during template validation.

Other variables can be defined globally with `NETJSONCONFIG_CONTEXT`, in that case we can simply pass those as they are defined.